pkg/execution/execution.go

```diff
@@ -21,57 +21,77 @@
 package execution
 
 import (
-	"context"
 	"errors"
 	"time"
 
 	"github.com/dop251/goja"
 
 	"go.k6.io/k6/js/common"
+	"go.k6.io/k6/js/modules"
 	"go.k6.io/k6/lib"
 )
 
-// Execution is a JS module to return information about the execution in progress.
-type Execution struct{}
+type (
+	// RootModule is the global module instance that will create module
+	// instances for each VU.
+	RootModule struct{}
 
-// New returns a pointer to a new Execution.
-func New() *Execution {
-	return &Execution{}
-}
-
-// GetVUStats returns information about the currently executing VU.
-func (e *Execution) GetVUStats(ctx context.Context) (goja.Value, error) {
-	vuState := lib.GetState(ctx)
-	if vuState == nil {
-		return nil, errors.New("getting VU information in the init context is not supported")
+	// ModuleInstance represents an instance of the execution module.
+	ModuleInstance struct {
+		modules.InstanceCore
+		obj *goja.Object
 	}
+)
 
-	rt := common.GetRuntime(ctx)
-	if rt == nil {
-		return nil, errors.New("goja runtime is nil in context")
-	}
+var (
+	_ modules.IsModuleV2 = &RootModule{}
+	_ modules.Instance   = &ModuleInstance{}
+)
 
-	stats := map[string]interface{}{
-		"id":        vuState.VUID,
-		"idGlobal":  vuState.VUIDGlobal,
-		"iteration": vuState.Iteration,
-		"iterationScenario": func() goja.Value {
-			return rt.ToValue(vuState.GetScenarioVUIter())
-		},
-	}
+// New returns a pointer to a new RootModule instance.
+func New() *RootModule {
+	return &RootModule{}
+}
 
-	obj, err := newLazyJSObject(rt, stats)
-	if err != nil {
-		return nil, err
+// NewModuleInstance implements the modules.IsModuleV2 interface to return
+// a new instance for each VU.
+func (*RootModule) NewModuleInstance(m modules.InstanceCore) modules.Instance {
+	mi := &ModuleInstance{InstanceCore: m}
+	rt := m.GetRuntime()
+	o := rt.NewObject()
+	defProp := func(name string, newInfo func() (*execInfo, error)) {
+		err := o.DefineAccessorProperty(name, rt.ToValue(func() goja.Value {
+			rt := mi.GetRuntime()
+			dobj, err := newInfo()
+			if err != nil {
+				common.Throw(rt, err)
+			}
+			return rt.NewDynamicObject(dobj)
```

Comment:
Is there a reason why we are using a DynamicObject here instead of again a simple goja.Object? I remember there was something but I can't remember what

Reply:
Hhmm don't think we had a specific reason, other than wanting the access to be lazy-loaded, but yeah, that can also be done with property accessors. So sure, I'll get rid of this as well.

Reply:
See 65326ec1.